lightning/src/events/mod.rs

```diff
@@ -2255,10 +2260,34 @@ impl MaybeReadable for Event {
 						(1, failure_reason, option),
 						(2, failure_type_opt, upgradable_required),
 					});
-					Ok(Some(Event::HTLCHandlingFailed {
+
+					let event = Event::HTLCHandlingFailed {
 						prev_channel_id,
 						failure_type: _init_tlv_based_struct_field!(failure_type_opt, upgradable_required),
 						failure_reason,
+					};
+
+					// If a legacy HTLCHandlingFailureType::UnknownNextHop was written, upgrade
+					// it to its new representation, otherwise leave unchanged.
+					Ok(Some(match event {
+						Event::HTLCHandlingFailed { prev_channel_id, ref failure_type, ref failure_reason } => {
```

Comment:
This set and match approach - wasn't it possible to work with the separate fields, do the transform if needed, and only at the very end build the Event struct and return?

Reply:
Played with this locally and I tend to agree, I like `if let` here to avoid a panic:

```rust
					// If a legacy HTLCHandlingFailureType::UnknownNextHop was written, upgrade
					// it to its new representation, otherwise leave unchanged.
					if let Some(HTLCHandlingFailureType::UnknownNextHop { requested_forward_scid }) = failure_type_opt.0 {
						failure_type_opt.0 = Some(HTLCHandlingFailureType::InvalidForward { requested_forward_scid });
						failure_reason = Some(LocalHTLCFailureReason::UnknownNextPeer.into());
					}
					Ok(Some(Event::HTLCHandlingFailed {
						prev_channel_id,
						failure_type: _init_tlv_based_struct_field!(failure_type_opt, upgradable_required),
						failure_reason
					}))

Reply:
Fixed in c952a7549eef052b71a7564949b63fb5c48cf3d5 👍